syntax = "proto3";

option go_package = "cactus/tango/src; protobuff";

package protobuff;

service TangoService {
  rpc SubmitTask(TaskRequest) returns (TaskResponse) {}
  rpc RegisterDevice(DeviceInfo) returns (DeviceResponse) {}
  rpc UpdateDeviceStatus(DeviceStatus) returns (DeviceStatusResponse) {}
  rpc FetchTask(DeviceRequest) returns (TaskAssignment) {}
  rpc ReportResult(TaskResult) returns (ResultResponse) {}
  rpc GetJobStatus(JobStatusRequest) returns (JobStatusReply) {}
}

message TaskRequest {
  string job_id = 1;
  string operation = 2;   
  bytes a_data = 3;   
  bytes b_data = 4;  
  optional bytes w_data = 5;
  optional bytes scale_bytes = 6; 
  optional float scale_scalar = 7; 
  int32 num_splits = 8;
  int32 m = 9;
  int32 n = 10;
  int32 d = 11;
}

message TaskResponse {
  bool accepted = 1;
  string message = 2;
}

message DeviceInfo {
  string device_id = 1;
  float internet_speed = 2;
  int32 available_ram = 3;
  int32 cpu_usage = 4;
  bool is_charging = 5;
}

message DeviceResponse {
  bool registered = 1;
  string message = 2;
}

message DeviceStatus {
  string device_id = 1;
  int32 tasks_in_last_hour = 2;
  float cpu_usage = 3;
  int32 available_ram = 4;
  float internet_speed = 5;
  bool is_charging = 6;
}

message DeviceStatusResponse {
  bool updated = 1;
  string message = 2;
}

message DeviceRequest {
  string device_id = 1;
}

message TaskAssignment {
  string job_id = 1;
  string task_id = 2;
  string operation = 3;   
  bytes a_data = 4;      
  bytes b_data = 5; 
  optional bytes w_data = 6;
  optional bytes scale_bytes = 7; 
  optional float scale_scalar = 8; 
  int32 num_splits = 9;
  int32 m = 10;
  int32 n = 11;
  int32 d = 12;
}

message TaskResult {
  string device_id = 1;
  string job_id = 2;
  string task_id = 3;
  bytes result_data = 4; 
}

message ResultResponse {
  bool success = 1;
  string message = 2;
}

message JobStatusRequest {
  string job_id = 1;
}

message JobStatusReply {
  bool is_complete = 1;
  string message = 2;
  bytes final_result = 3;
}
